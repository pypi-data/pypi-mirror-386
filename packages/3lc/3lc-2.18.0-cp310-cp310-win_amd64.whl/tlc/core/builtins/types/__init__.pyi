from .bb_crop_interface import BBCropInterface as BBCropInterface
from typing import Any, TypedDict

SampleData = Any
MetricData = Any

class MetricTableInfo(TypedDict):
    """A dictionary containing summary metadata about a metric table."""
    url: str
    file_size: int
    stream_name: str
    row_count: int
