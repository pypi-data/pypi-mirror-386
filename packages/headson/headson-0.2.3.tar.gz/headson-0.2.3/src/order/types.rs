use serde_json;

#[derive(Clone, Debug)]
pub struct PriorityConfig {
    pub max_string_graphemes: usize,
    pub array_max_items: usize,
}

impl PriorityConfig {
    pub fn new(max_string_graphemes: usize, array_max_items: usize) -> Self {
        Self {
            max_string_graphemes,
            array_max_items,
        }
    }
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub struct NodeId(pub usize);

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum NodeKind {
    Null,
    Bool,
    Number,
    String,
    Array,
    Object,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum ObjectType {
    Object,
    Fileset,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct RankedNode {
    pub node_id: NodeId,
    pub kind: NodeKind,
    pub key_in_object: Option<String>,
    pub number_value: Option<serde_json::Number>,
    pub bool_value: Option<bool>,
    pub string_value: Option<String>,
}

#[derive(Clone, Debug, Default)]
pub struct NodeMetrics {
    pub array_len: Option<usize>,
    pub object_len: Option<usize>,
    pub string_len: Option<usize>,
    pub string_truncated: bool,
}

#[derive(Clone, Debug)]
pub struct PriorityOrder {
    pub metrics: Vec<NodeMetrics>,
    pub nodes: Vec<RankedNode>,
    // All ids in this structure are PQ ids (0..total_nodes).
    // They correspond to `NodeId.0` in `RankedNode` for convenience when indexing.
    pub parent: Vec<Option<NodeId>>, // parent[id] = parent id (PQ id)
    pub children: Vec<Vec<NodeId>>,  // children[id] = children ids (PQ ids)
    pub order: Vec<NodeId>, // ids sorted by ascending priority (PQ ids)
    pub total_nodes: usize,
    pub object_type: Vec<ObjectType>,
}

pub const ROOT_PQ_ID: usize = 0;
