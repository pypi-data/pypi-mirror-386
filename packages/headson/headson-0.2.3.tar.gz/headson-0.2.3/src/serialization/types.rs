#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum OutputTemplate {
    Json,
    Pseudo,
    Js,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RenderConfig {
    pub template: OutputTemplate,
    pub indent_unit: String,
    pub space: String,
    // Newline sequence to use in final output (e.g., "\n" or "").
    // Templates read this directly; no post-processing replacement.
    pub newline: String,
}
