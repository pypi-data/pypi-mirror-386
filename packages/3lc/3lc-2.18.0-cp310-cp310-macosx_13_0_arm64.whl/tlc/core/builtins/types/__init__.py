# Pyarmor 9.1.9 (pro), 005329, 3lc, 2025-10-22T14:50:26.339468
from tlcsaas.pyarmor_runtime_005329 import __pyarmor__
__pyarmor__(__name__, __file__, b'PY005329\x00\x03\n\x00o\r\r\n\x80\x00\x01\x00\x08\x00\x00\x00\x04\x00\x00\x00@\x00\x00\x00e\x04\x00\x00\x12\t\x06\x00\xd0H\x15\xba\xc5\x01d\x8e\xe6c\x19\x82`\x9a3\xc8\x00\x00\x00\x00\x00\x00\x00\x00,\x88\xb4\x05B.\xff\x00\xe9\xe29\xaa\x84\xa7O\xbb\x13\x0c\x0b\xde\xc6s-\xf9\x99;\xb3%{\xe5Q\x06\xc7\x9a{\xea\xa7M}\xb7\xb1\xf5\xb3\x8al\x89\xcb\xdc9\x86\x1c\x8ak\xdfx\xd9\x96\xd9L\x8b\x9d\xdf\xe5\xdbf\x1c\xa8w\xff\xca\x89U\xa6\xfed\xf0\xe9R\x11\x9d3\xe7\xf9\x01k@\x1b \x81l\x84\xb1_qX\xfe\x90\x0b\x97D<o\xfa\x10\xd2\xa0\xe1,|\xa5\xf3fXY\x19\xce\xd6>d\xaf\xde{8\xb8U3y\r\xb2\xd0"{\xaf:\xca\x9cJ\x94k\x86\xa6\xd5o\xad\xa0\xfa?R\n\x8a\x12\xed\x84n\x03v\xa7\xd4Ln\xcb,nFm\x1e\xde\xee\xbcf\xaa\xbe\xfd/\xe2\x86]\x0c\x9a\x02D\xaf\xe3\x96vO-\xcb\xf0\xb1]\x05h\xdb\x99\x8f~\xb4\xf5z\xde\x8d\x18\x89O;~\xe76rd\xa9B\x8bX\xc2\x8eZ7\xcct\x9d\xfe\n\xcaO~77|Q\xe8\xaa\xb0"\n\xbc;v\xa6\x8aC7\xb2\xa1\xcbb\xc4\x11\xa7\xd3Cg\xe2|\xd8\xf4\xdc\xd2z+\xb5\xb7\x8a`\xb9t/U\x17\xa7\xdf\x97@a\xa5B\x98\x0b!\x0e1\xdc\xa8Eu|\x8b\x10\xa7\xf7N\x87\xf4\x18\xcf>\x11\xd7\x93YJd\xbf\x81G\xb9\x0e6~\xfa\x92{T\xc9H^\xaa\x99\x8e\xda*=\xad\x9c\xfb#\x91th\x0b\x01\x9d\xcb\xf2\xde\xea7\x9e}OF\xb0\xa5\x0c)\xb1\xa1\x86\rv\x9f\x8eF}\x89\x1e\xe7\xb4\x14\xc0\xfb\x96pZm\x88\x8cCQ\x99\xaa\x94_\xec\xb4\xe5*p\xe6s\x9a\x8c\xeb\xb1d\x82\x945u#J\xc2\xbb\xbe\xb2\x87\x0eA\xd81\xde\xecr\xbd\x11Y:\xaf\xaa\x81J\xe8\xf0\xd2\x9da\x96\x9e\x07ok\x1e\r\x82\xa0\x07\x9c\x1d\x03i\xb4\xc1/\xa8$)e\xff\x8f\x0c\xbaY\x82\x1eS\xfe3&\x13Hp\xbd\xe2\x8dM\xc5\r\xa1\x0bWM\xb4q]a\x86\x03Z{\n\xf88\x8acNB9Y\x0f\xd6!\xd8HE\xeb4\xd4\xc4\x15\x1cb\xc7\x9d\xb7/\xb6\xf8\xe5\x8e^\x12\xcd,\xca@\x1aZz\x95XF\x13\xebd,\xc2\x81\xec\x18\xbf\xb6\x83E\xa3\xf3\xa9\xba\x86\xbeQE\xb5\x03\x05\x15\x11\xb7\xa7\xca\xbc@\xed\x9dm\xa2\t\xa90\xaf\xe7s\x03\x06\xb0d\x05\xe4\xe1~A\xf3\xc0\x03\xce#\xc1\x01s\xce8\x8d\x81\x07|?\xd19\xdb4\xf2\r\x0bt\xe3\xe8\xd5\x97\xcf\xde\r}?\x8b\xec\xa0R\x1b\x98\x0em\'N\xcf\xef\xc5&\xa6s}\xe9\x17\xdc6\xd1*0\xa3\xd9\x07\xd9\xd8\x81a\xa1\x14\xdb\x8d*l\xbf\xea\xf55o\xea30s\xf0e\x9e\x04-e\xea\x99\x11/B\x9b\x11?U\x95$\xba\xad\xc0\xd2L\x93\xee\xe0\x83d\xf4&`|J\xbd\x97O\xf2o\x17\xc6a\xc3\xf5d\xb0\xc5|\xd2\x86\x00\xeef!\x9f+\xc5\xd6\xb3\x08@z\xb9\xa0\xc2|\x17\x98\x8b\xc36\'~\xb7P\xa2r6\x1d&.\xdfO\xa1\xe5\t\x8c\xdc\xea\xbb\xbd\xe4\x88\xb1V&i\xf6\xea\x9a0\xbbz\x9d\xd7\'\xe2\xc0w#K\xef\xd0\x9dX\xd9^\x85\x0b\xefF\xba+t\x17I^\x04\x00t\xcc\xa2^\xe8W\xd7\x99\xf9\xb3L\x11\x01U[I\xac \x82\xfbp\x8dm\xa2\xfdy\xd8)!\x93\xe9\xb8\xdbx\x06f\x8ea\xf2\x98\xcc9LL\xa2\xb2\xc8\x1b\xf3\xb74\xf0\xaf\xd7nH\xe1\xacC?\xb2\xb2\xfb\x8f\x9a\xd74\xac&\xc0\xad\xd0b4X\x0f1\xf1\xa2\x94\x98\xf8*\xf3RHR\xfeW\n\xfe\xe3\x83K\xa2P\xc9\xe4\xe0\xefV9M\xb3"\xb5\nL\xf9C%wyO\xbdm\xc0\xa2\x1c\xed\xc0Od\x7f\xf2\x19\x19\xc0\xb3\xe1\xfej\x84\xb3\xbe\x85\x15\xd5,\xa5\xc0\xa3\x15\xf7Q\xb3\xb1\\dy\xb3\x1e\x8c\x10\xfc\xa2\x11K\xd4B\xc4\x96\xe1\x8c\x84\xa7|Y\xf8Z\x85\xc0E\x1e\xdf\x90\x97\xa9\x13\xdb\\\xdbt.\xb6\x9eAIg\x97\xbc.\xe7(\xcf\x9a\xd08\xee\xfd\x8d\xaa\x1a\xf4o\xfc\xc26\xd8\xd6\xed\x02\xb1?\xe1\xdf\xfe R\xb17\'f\xf0X\xaa\xa7G/9\xfc\xfc\xbbX\xafZ\xce\xb5c8J\x14V\xbdb\t\xeb\xa0\xadW0\xd6\xbc#\xccx\xbeWo\xb1\xd8\t\xb9\x9eT\xebd}F\x7f[\x16\xbfz\xd3\xb6\x0b\xe4m\xca\xe5\x01Y\x86\n\xcc\x0ed\x04\x96&\x8a\xc8\x93I3\xcd-L\xa0y\xd9\xbb\xf9\xe8\xd5w\xe3K\xb2*\x05\xaac(\xd3\xefO\xb5\x85\xd8\x9c\xf6?e\xa0\xd3IJ\x98Ayf(%\xc2\xc4W\x01vD*Wq\xe7\xaf6l{"\x96\xae\x113\xbf0$T\xc1')
