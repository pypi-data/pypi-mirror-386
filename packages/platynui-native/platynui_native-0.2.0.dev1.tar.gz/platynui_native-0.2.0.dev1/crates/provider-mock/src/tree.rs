use crate::focus;
use crate::input;
use crate::node::{MockNode, NodePatternContext, attr};
use crate::window;
use platynui_core::provider::ProviderDescriptor;
use platynui_core::types::{Point, Rect};
use platynui_core::ui::attribute_names::{activation_target, element, focusable};
use platynui_core::ui::{
    FocusableAction, Namespace, PatternId, PatternRegistry, RuntimeId, UiAttribute, UiNode, UiPattern, UiValue,
};
use quick_xml::de::from_str;
use serde::Deserialize;
use std::collections::HashMap;
use std::fmt;
use std::sync::{Arc, LazyLock, RwLock};

type NodeList = Vec<Arc<MockNode>>;
pub(crate) type ProviderTree = (NodeList, NodeList, HashMap<String, Arc<MockNode>>);

static CURRENT_TREE: LazyLock<RwLock<StaticMockTree>> = LazyLock::new(|| RwLock::new(StaticMockTree::default()));

type InstantiatedTree = (NodeList, NodeList, NodeList);

#[derive(Clone, Debug)]
pub struct AttributeSpec {
    namespace: Namespace,
    name: String,
    value: UiValue,
}

impl AttributeSpec {
    pub fn new(namespace: Namespace, name: impl Into<String>, value: UiValue) -> Self {
        Self { namespace, name: name.into(), value }
    }

    pub fn namespace(&self) -> Namespace {
        self.namespace
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn value(&self) -> &UiValue {
        &self.value
    }
}

impl From<(Namespace, &'static str, UiValue)> for AttributeSpec {
    fn from(value: (Namespace, &'static str, UiValue)) -> Self {
        AttributeSpec::new(value.0, value.1, value.2)
    }
}

impl From<(Namespace, String, UiValue)> for AttributeSpec {
    fn from(value: (Namespace, String, UiValue)) -> Self {
        AttributeSpec::new(value.0, value.1, value.2)
    }
}

#[derive(Clone, Debug)]
pub struct NodeSpec {
    namespace: Namespace,
    role: String,
    name: String,
    runtime_id: String,
    attributes: Vec<AttributeSpec>,
    patterns: Vec<String>,
    children: Vec<NodeSpec>,
    expose_flat: bool,
    order_key: Option<u64>,
    text: Option<String>,
}

impl NodeSpec {
    pub fn new(
        namespace: Namespace,
        role: impl Into<String>,
        name: impl Into<String>,
        runtime_id: impl Into<String>,
    ) -> Self {
        Self {
            namespace,
            role: role.into(),
            name: name.into(),
            runtime_id: runtime_id.into(),
            attributes: Vec::new(),
            patterns: Vec::new(),
            children: Vec::new(),
            expose_flat: false,
            order_key: None,
            text: None,
        }
    }

    pub fn with_attribute(mut self, attribute: impl Into<AttributeSpec>) -> Self {
        self.attributes.push(attribute.into());
        self
    }

    pub fn with_child(mut self, child: NodeSpec) -> Self {
        self.children.push(child);
        self
    }

    pub fn with_pattern(mut self, pattern: impl Into<String>) -> Self {
        self.patterns.push(pattern.into());
        self
    }

    pub fn with_patterns<I, S>(mut self, patterns: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.patterns.extend(patterns.into_iter().map(Into::into));
        self
    }

    pub fn with_text(mut self, text: impl Into<String>) -> Self {
        self.text = Some(text.into());
        self
    }

    pub fn with_expose_flat(mut self, expose: bool) -> Self {
        self.expose_flat = expose;
        self
    }

    pub fn expose_flat(&self) -> bool {
        self.expose_flat
    }
}

fn collect_flat_specs(spec: &NodeSpec, flat_specs: &mut Vec<NodeSpec>) {
    if spec.expose_flat {
        flat_specs.push(clone_for_flat(spec));
    }
    for child in &spec.children {
        collect_flat_specs(child, flat_specs);
    }
}

fn clone_for_flat(spec: &NodeSpec) -> NodeSpec {
    let mut clone = spec.clone();
    clone.expose_flat = false;
    clone.order_key = None;
    clone.children = spec.children.iter().map(clone_for_flat).collect();
    clone
}

fn assign_order_keys(spec: &mut NodeSpec, counter: &mut u64) {
    spec.order_key = Some(*counter);
    *counter += 1;
    for child in &mut spec.children {
        assign_order_keys(child, counter);
    }
}

#[derive(Clone, Debug)]
pub struct StaticMockTree {
    roots: Vec<NodeSpec>,
    flat_specs: Vec<NodeSpec>,
}

impl StaticMockTree {
    pub fn new(roots: Vec<NodeSpec>) -> Self {
        let mut flat_specs = Vec::new();
        for spec in &roots {
            collect_flat_specs(spec, &mut flat_specs);
        }

        let mut roots = roots;
        let mut counter = 0u64;
        for spec in &mut roots {
            assign_order_keys(spec, &mut counter);
        }
        for spec in &mut flat_specs {
            assign_order_keys(spec, &mut counter);
        }

        Self { roots, flat_specs }
    }

    pub fn roots(&self) -> &[NodeSpec] {
        &self.roots
    }

    pub fn flat_specs(&self) -> &[NodeSpec] {
        &self.flat_specs
    }

    fn instantiate(&self, descriptor: &ProviderDescriptor) -> InstantiatedTree {
        let mut all = Vec::new();
        let roots: NodeList =
            self.roots.iter().map(|spec| instantiate_node(spec, descriptor, None, &mut all)).collect();
        let flat: NodeList =
            self.flat_specs.iter().map(|spec| instantiate_node(spec, descriptor, None, &mut all)).collect();
        (roots, flat, all)
    }
}

impl Default for StaticMockTree {
    fn default() -> Self {
        const XML: &str = include_str!("../assets/mock_tree.xml");
        Self::from_xml(XML).expect("embedded mock_tree.xml could not be parsed")
    }
}

impl StaticMockTree {
    pub fn from_xml(xml: &str) -> Result<Self, MockTreeLoadError> {
        let parsed: XmlTree = from_str(xml).map_err(MockTreeLoadError::Xml)?;
        let mut roots = Vec::new();
        for node in parsed.nodes {
            roots.push(build_node(node)?);
        }
        Ok(StaticMockTree::new(roots))
    }
}

#[derive(Debug, Deserialize)]
struct XmlTree {
    #[serde(rename = "node", default)]
    nodes: Vec<XmlNode>,
}

#[derive(Debug, Deserialize)]
struct XmlNode {
    #[serde(rename = "@namespace")]
    namespace: String,
    #[serde(rename = "@role")]
    role: String,
    #[serde(rename = "@name")]
    name: String,
    #[serde(rename = "@runtime_id")]
    runtime_id: String,
    #[serde(rename = "@bounds")]
    bounds: Option<String>,
    #[serde(rename = "@activation_point")]
    activation_point: Option<String>,
    #[serde(rename = "@visible")]
    visible: Option<bool>,
    #[serde(rename = "@enabled")]
    enabled: Option<bool>,
    #[serde(rename = "@expose_flat")]
    expose_flat: Option<bool>,
    #[serde(rename = "attribute", default)]
    attributes: Vec<XmlAttribute>,
    #[serde(rename = "patterns")]
    patterns: Option<XmlPatternList>,
    #[serde(rename = "text")]
    text: Option<String>,
    #[serde(rename = "node", default)]
    children: Vec<XmlNode>,
}

#[derive(Debug, Deserialize)]
struct XmlAttribute {
    #[serde(rename = "@namespace")]
    namespace: Option<String>,
    #[serde(rename = "@name")]
    name: String,
    #[serde(rename = "@value")]
    value: String,
}

#[derive(Debug, Deserialize)]
struct XmlPatternList {
    #[serde(rename = "$text")]
    value: Option<String>,
}

fn build_node(node: XmlNode) -> Result<NodeSpec, MockTreeLoadError> {
    let namespace = parse_namespace(&node.namespace)?;
    let mut spec = NodeSpec::new(namespace, node.role.clone(), node.name.clone(), node.runtime_id.clone());

    if let Some(patterns) = parse_patterns(node.patterns.as_ref()) {
        spec.patterns.extend(patterns);
    }

    let visible = node.visible.unwrap_or(true);
    let enabled = node.enabled.unwrap_or(true);

    if let Some(bounds) = node.bounds.as_ref() {
        let rect = parse_rect(bounds)?;
        push_bounds_attributes(&mut spec, namespace, rect, visible, enabled);
    } else {
        push_visibility_attributes(&mut spec, namespace, visible, enabled);
    }

    if let Some(text) = node.text.as_ref() {
        spec.text = Some(text.clone());
    }

    if let Some(point_str) = node.activation_point.as_ref() {
        let point = parse_point(point_str)?;
        push_activation_point_attributes(&mut spec, namespace, point);
    }

    for attr in node.attributes {
        let attr_namespace = match attr.namespace {
            Some(ns) => parse_namespace(&ns)?,
            None => namespace,
        };
        spec.attributes.push(AttributeSpec::new(attr_namespace, attr.name, parse_attribute_value(&attr.value)));
    }

    spec.expose_flat = node.expose_flat.unwrap_or(false);
    for child in node.children {
        spec.children.push(build_node(child)?);
    }

    Ok(spec)
}

fn parse_namespace(value: &str) -> Result<Namespace, MockTreeLoadError> {
    match value {
        "control" => Ok(Namespace::Control),
        "item" => Ok(Namespace::Item),
        "app" => Ok(Namespace::App),
        "native" => Ok(Namespace::Native),
        other => Err(MockTreeLoadError::UnknownNamespace(other.to_owned())),
    }
}

fn parse_rect(value: &str) -> Result<Rect, MockTreeLoadError> {
    let parts: Vec<f64> = value
        .split(',')
        .map(|chunk| chunk.trim().parse::<f64>())
        .collect::<Result<_, _>>()
        .map_err(|_| MockTreeLoadError::InvalidRect(value.to_owned()))?;
    if parts.len() != 4 {
        return Err(MockTreeLoadError::InvalidRect(value.to_owned()));
    }
    Ok(Rect::new(parts[0], parts[1], parts[2], parts[3]))
}

fn parse_point(value: &str) -> Result<Point, MockTreeLoadError> {
    let parts: Vec<f64> = value
        .split(',')
        .map(|chunk| chunk.trim().parse::<f64>())
        .collect::<Result<_, _>>()
        .map_err(|_| MockTreeLoadError::InvalidPoint(value.to_owned()))?;
    if parts.len() != 2 {
        return Err(MockTreeLoadError::InvalidPoint(value.to_owned()));
    }
    Ok(Point::new(parts[0], parts[1]))
}

fn parse_attribute_value(value: &str) -> UiValue {
    if let Ok(boolean) = value.parse::<bool>() {
        return UiValue::from(boolean);
    }
    if let Ok(integer) = value.parse::<i64>() {
        return UiValue::from(integer);
    }
    if let Ok(number) = value.parse::<f64>() {
        return UiValue::from(number);
    }
    UiValue::from(value.to_owned())
}

fn parse_patterns(list: Option<&XmlPatternList>) -> Option<Vec<String>> {
    let raw = list.and_then(|p| p.value.as_ref())?;
    let entries = raw
        .split(',')
        .map(|entry| entry.trim())
        .filter(|entry| !entry.is_empty())
        .map(|entry| entry.to_owned())
        .collect::<Vec<_>>();
    if entries.is_empty() { None } else { Some(entries) }
}

fn push_bounds_attributes(spec: &mut NodeSpec, namespace: Namespace, rect: Rect, visible: bool, enabled: bool) {
    push_visibility_attributes(spec, namespace, visible, enabled);
    spec.attributes.push(AttributeSpec::new(namespace, element::BOUNDS, UiValue::Rect(rect)));
}

fn push_visibility_attributes(spec: &mut NodeSpec, namespace: Namespace, visible: bool, enabled: bool) {
    spec.attributes.push(AttributeSpec::new(namespace, element::IS_VISIBLE, UiValue::from(visible)));
    spec.attributes.push(AttributeSpec::new(namespace, element::IS_ENABLED, UiValue::from(enabled)));
}

fn push_activation_point_attributes(spec: &mut NodeSpec, namespace: Namespace, point: Point) {
    spec.attributes.push(AttributeSpec::new(namespace, activation_target::ACTIVATION_POINT, UiValue::Point(point)));
}

#[derive(Debug)]
pub enum MockTreeLoadError {
    Xml(quick_xml::DeError),
    UnknownNamespace(String),
    InvalidRect(String),
    InvalidPoint(String),
}

impl fmt::Display for MockTreeLoadError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MockTreeLoadError::Xml(err) => write!(f, "XML parsing failed: {err}"),
            MockTreeLoadError::UnknownNamespace(ns) => {
                write!(f, "unknown namespace '{ns}' in mock tree")
            }
            MockTreeLoadError::InvalidRect(value) => {
                write!(f, "invalid bounds format '{value}' (expected x,y,width,height)")
            }
            MockTreeLoadError::InvalidPoint(value) => {
                write!(f, "invalid point format '{value}' (expected x,y)")
            }
        }
    }
}

impl std::error::Error for MockTreeLoadError {}

impl From<quick_xml::DeError> for MockTreeLoadError {
    fn from(err: quick_xml::DeError) -> Self {
        MockTreeLoadError::Xml(err)
    }
}

pub struct TreeGuard {
    previous: StaticMockTree,
}

impl Drop for TreeGuard {
    fn drop(&mut self) {
        *CURRENT_TREE.write().unwrap() = self.previous.clone();
    }
}

pub fn install_mock_tree(tree: StaticMockTree) -> TreeGuard {
    let mut lock = CURRENT_TREE.write().unwrap();
    let previous = lock.clone();
    *lock = tree;
    TreeGuard { previous }
}

pub fn reset_mock_tree() {
    *CURRENT_TREE.write().unwrap() = StaticMockTree::default();
}

pub(crate) fn instantiate_nodes(descriptor: &ProviderDescriptor) -> ProviderTree {
    input::reset_text_buffers();
    let tree = CURRENT_TREE.read().unwrap().clone();
    let (roots, flat_nodes, all_nodes) = tree.instantiate(descriptor);
    let mut map = HashMap::new();
    for node in all_nodes {
        map.insert(node.runtime_id().as_str().to_owned(), node.clone());
    }
    (roots, flat_nodes, map)
}

fn instantiate_node(
    spec: &NodeSpec,
    descriptor: &ProviderDescriptor,
    parent: Option<&Arc<MockNode>>,
    all: &mut Vec<Arc<MockNode>>,
) -> Arc<MockNode> {
    let runtime_id = RuntimeId::from(spec.runtime_id.as_str());

    let runtime_patterns = PatternRegistry::new();
    let mut dynamic_attributes: Vec<Arc<dyn UiAttribute>> = Vec::new();
    let mut declared_patterns: Vec<PatternId> = Vec::new();

    if let Some(text) = spec.text.as_ref() {
        dynamic_attributes.push(input::register_text_attribute(spec.namespace, &runtime_id, text));
    }

    let has_window_surface = spec.patterns.iter().any(|pattern| pattern == "WindowSurface");
    let has_focusable = spec.patterns.iter().any(|pattern| pattern == "Focusable");
    let window_config = has_window_surface.then(|| window::derive_config(&spec.attributes));

    let initial_focus = if has_focusable {
        spec.attributes.iter().any(|attr| {
            attr.namespace() == Namespace::Control
                && attr.name() == focusable::IS_FOCUSED
                && matches!(attr.value(), UiValue::Bool(true))
        })
    } else {
        false
    };

    for pattern in &spec.patterns {
        match pattern.as_str() {
            "Focusable" => {
                let action_runtime_id = runtime_id.clone();
                runtime_patterns.register_lazy(PatternId::from("Focusable"), move || {
                    let target = action_runtime_id.clone();
                    let pattern: Arc<dyn UiPattern> =
                        Arc::new(FocusableAction::new(move || focus::request_focus(target.clone())));
                    Some(pattern)
                });
                dynamic_attributes.push(focus::focus_attribute(spec.namespace, runtime_id.clone()));
            }
            "WindowSurface" => {}
            other => declared_patterns.push(PatternId::from(other)),
        }
    }

    if let Some(config) = window_config {
        dynamic_attributes.extend(window::register_window(
            runtime_id.clone(),
            spec.namespace,
            config,
            &runtime_patterns,
        ));
    }

    if has_focusable && initial_focus {
        let _ = focus::request_focus(runtime_id.clone());
    }

    let pattern_context = NodePatternContext { runtime_patterns, declared_patterns, order_key: spec.order_key };

    let technology = descriptor.technology.as_str();

    let mut attributes: Vec<Arc<dyn UiAttribute>> = spec
        .attributes
        .iter()
        .filter(|attr_spec| {
            if has_window_surface && window::should_filter_attribute(attr_spec.name()) {
                return false;
            }
            if has_focusable && attr_spec.namespace() == Namespace::Control && attr_spec.name() == focusable::IS_FOCUSED
            {
                return false;
            }
            true
        })
        .map(|attr_spec| attr(attr_spec.namespace(), attr_spec.name().to_owned(), attr_spec.value().clone()))
        .collect();
    attributes.extend(dynamic_attributes);

    let node = MockNode::new(
        spec.namespace,
        spec.role.clone(),
        spec.name.clone(),
        runtime_id.as_str(),
        technology,
        attributes,
        pattern_context,
    );

    if let Some(parent_node) = parent {
        MockNode::add_child(parent_node, Arc::clone(&node));
    }

    all.push(Arc::clone(&node));

    for child in &spec.children {
        instantiate_node(child, descriptor, Some(&node), all);
    }

    node
}
