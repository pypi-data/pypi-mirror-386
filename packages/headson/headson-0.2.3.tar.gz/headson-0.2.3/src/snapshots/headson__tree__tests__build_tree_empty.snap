---
source: src/tree.rs
expression: "tree.serialize(OutputTemplate::Json)"
---
[]
